### DISCUSSION TITLE
Error in FastAPI documentation? Self-hosting JavaScript and CSS for docs doesn't work

### DISCUSSION THREAD

**Discussion by imthematrix** [fastapi/fastapi#11148] [Questions]
<details> 
  <summary>First Check</summary>

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
</details>

<details> 
  <summary>Commit to Help</summary>

- [X] I commit to help with one of those options 👆
</details>

<br>

I think there's an error in the FastAPI documentation for [Self-hosting JavaScript and CSS for docs](https://fastapi.tiangolo.com/how-to/custom-docs-ui-assets/#self-hosting-javascript-and-css-for-docs). I copied the file structure and code from the user guide and followed the tutorial exactly but it doesn't work.

### Example Code

This is my code. It is completely copied from the FastAPI documentation [Self-hosting JavaScript and CSS for docs](https://fastapi.tiangolo.com/how-to/custom-docs-ui-assets/#self-hosting-javascript-and-css-for-docs) and **not modified**:

<details> 
  <summary>main.py</summary>

```python
from fastapi import FastAPI
from fastapi.openapi.docs import (
    get_redoc_html,
    get_swagger_ui_html,
    get_swagger_ui_oauth2_redirect_html,
)
from fastapi.staticfiles import StaticFiles

app = FastAPI(docs_url=None, redoc_url=None)

app.mount("/static", StaticFiles(directory="static"), name="static")


@app.get("/docs", include_in_schema=False)
async def custom_swagger_ui_html():
    return get_swagger_ui_html(
        openapi_url=app.openapi_url,
        title=app.title + " - Swagger UI",
        oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
        swagger_js_url="/static/swagger-ui-bundle.js",
        swagger_css_url="/static/swagger-ui.css",
    )


@app.get(app.swagger_ui_oauth2_redirect_url, include_in_schema=False)
async def swagger_ui_redirect():
    return get_swagger_ui_oauth2_redirect_html()


@app.get("/redoc", include_in_schema=False)
async def redoc_html():
    return get_redoc_html(
        openapi_url=app.openapi_url,
        title=app.title + " - ReDoc",
        redoc_js_url="/static/redoc.standalone.js",
    )


@app.get("/users/{username}")
async def read_user(username: str):
    return {"message": f"Hello {username}"}
```
</details>

My file structure is also copied exactly from the docs. My file structure looks like this:
```
C:\PycharmProjects\FastAPI Test
├── app
│   ├── __init__.py
│   ├── main.py
├── static
│   ├── redoc.standalone.js
│   ├── swagger-ui-bundle.js
│   └── swagger-ui.css
├── poetry.lock
└── pyproject.toml
```

The file structure in the docs looks like this:
```
.
├── app
│   ├── __init__.py
│   ├── main.py
└── static
    ├── redoc.standalone.js
    ├── swagger-ui-bundle.js
    └── swagger-ui.css
```

### Description

I expect FastAPI to work and show the docs when I run `C:\PycharmProjects\FastAPI Test\app> uvicorn main:app --reload` but it **doesn't work**. I receive this error: `RuntimeError: Directory 'static' does not exist`.

FastAPI doesn't find the `static` folder.

<details> 
  <summary>Console log</summary>

```
(fastapi test 12-py3.11) PS C:\PycharmProjects\FastAPI Test\app> uvicorn main:app --reload
INFO:     Will watch for changes in these directories: ['H:\\PycharmProjects\\FastAPI Test 12\\app']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [46164] using WatchFiles
Process SpawnProcess-1:
Traceback (most recent call last):
  File "C:\Program Files\Python311\Lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
```
  ...
```
  File "C:\PycharmProjects\FastAPI Test\app\main.py", line 11, in <module>
    app.mount("/static", StaticFiles(directory="static"), name="static")
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\pypoetry\Cache\virtualenvs\fastapi_test_12-9IkIsdSL-py3.11\Lib\site-packages\starlette\staticfiles.py", line 59, in __init__
    raise RuntimeError(f"Directory '{directory}' does not exist")
RuntimeError: Directory 'static' does not exist
```

(Traceback is shortened)
</details>

### Solution

It works when I change my file structure like this and run the same command in the `FastAPI Test` directory: `C:\PycharmProjects\FastAPI Test> uvicorn main:app --reload`
```
C:\PycharmProjects\FastAPI Test
├── main.py
├── static
│   ├── redoc.standalone.js
│   ├── swagger-ui-bundle.js
│   └── swagger-ui.css
├── poetry.lock
└── pyproject.toml
```

### Conclusion

Is the documentation wrong? When the `static` directory is in the same directory as the `main.py`, the code works fine. But it doesn't work as it is shown in the documentation (when `main` is in another directory).

<details> 
  <summary>Environment</summary>

### Operating System

Windows

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

Python 3.11.0

</details>

**Comment by Enescigdem** [comment]
Hi , Any solution ?

**Comment by luebke-dev** [comment]
You should use the absolut path of the static directory 

`StaticFiles(directory="/path/to/my/static")`

In your code the path is relative to the workdir

**Reply by Enescigdem** [reply]
Unfortunately , it did not work for me

**Comment by hasansezertasan** [comment]
Can you try this code block:

```python
from fastapi import FastAPI
from fastapi.openapi.docs import (
    get_redoc_html,
    get_swagger_ui_html,
    get_swagger_ui_oauth2_redirect_html,
)
import os
from fastapi.staticfiles import StaticFiles

app = FastAPI(docs_url=None, redoc_url=None)
statics_dir = os.path.join(os.path.dirname(__file__), "static")
app.mount("/static", StaticFiles(directory=statics_dir), name="static")


@app.get("/docs", include_in_schema=False)
async def custom_swagger_ui_html():
    return get_swagger_ui_html(
        openapi_url=app.openapi_url,
        title=app.title + " - Swagger UI",
        oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
        swagger_js_url="/static/swagger-ui-bundle.js",
        swagger_css_url="/static/swagger-ui.css",
    )


@app.get(app.swagger_ui_oauth2_redirect_url, include_in_schema=False)
async def swagger_ui_redirect():
    return get_swagger_ui_oauth2_redirect_html()


@app.get("/redoc", include_in_schema=False)
async def redoc_html():
    return get_redoc_html(
        openapi_url=app.openapi_url,
        title=app.title + " - ReDoc",
        redoc_js_url="/static/redoc.standalone.js",
    )


@app.get("/users/{username}")
async def read_user(username: str):
    return {"message": f"Hello {username}"}

```

**Reply by Enescigdem** [reply]
It did not work too . I use proxy with "root_path" variable in FastAPI . Can this "root_path" cause a problem ?

**Reply by Enescigdem** [reply]
When root_path is given , it gives  404 not found error

**Reply by hasansezertasan** [reply]
Can you check the network panel on the browser devtools and share the screenshot?

**Reply by Enescigdem** [reply]
![image](https://github.com/tiangolo/fastapi/assets/31999546/971f05d3-63e9-48f5-afe7-e97bfe086566)

**Reply by hasansezertasan** [reply]
I see, the script is not loaded. I'm not sure if it's caused by `root_path` or not.

**Reply by Enescigdem** [reply]
It works well unless root_path is given

**Reply by CastIronFace** [reply]
I just noticed this as well. If a root_path is given the root_path is duplicated for my static files. It looks like /{root_path}/{root_path}/static

**Reply by CastIronFace** [reply]
A little more information. The downgrading to version 0.108.0 fixed this issue for me.
